src/compiler/checker.ts

```diff
@@ -17018,9 +17019,9 @@ namespace ts {
         }
 
         function getRestrictiveTypeParameter(tp: TypeParameter) {
-            return tp.constraint === unknownType ? tp : tp.restrictiveInstantiation || (
+            return !tp.constraint && !getConstraintDeclaration(tp) || tp.constraint === noConstraintType ? tp : tp.restrictiveInstantiation || (
                 tp.restrictiveInstantiation = createTypeParameter(tp.symbol),
-                (tp.restrictiveInstantiation as TypeParameter).constraint = unknownType,
+                (tp.restrictiveInstantiation as TypeParameter).constraint = noConstraintType,

```

Comment:
I think you mentioned it in person, but remind me again: Why does this need to be `noConstraintType`? `AFAIK, `noConstraintType` should mostly just be used as a marker to make `getBaseConstraintOfType` return `undefined`, but a type parameter constrained to `unknown` and a type parameter without a constraint are supposed to be functionally identical nowadays. So if this change was required somewhere, I think it implies something is subtly broken. In any case, we should probably make `noConstraintType` an `unknown` variant at some point (since it's still an empty object), even if it's never supposed to actually be used as a type.

Reply:
The one difference I an into is in `getConstraintOfDistributiveConditionalType`. It does nothing for type parameters declared without a constraint, but applies a distributive conditional type to a declared type parameter constraint, even if that constraint is `unknown`. Changing this would definitely be a breaking change.

Making matters worse, for a restrictive type parameter we used to jam in an `unknown` constraint even for a type parameter declared without a constraint--which effectively changed the observed constraint for a distributive conditional type.

Reply:
Ew, icky. Distributive constraints behaving differently for `T` vs `T extends unknown` is definitely pretty bad - do we know which direction's behavior is better? Given the direction of this PR, I'm led to think we should adopt the unconstrained `T` behavior for a `T extends unknown`.